Add Bouhadjar sequence learning network tutorial

clinssen opened the pull request:
Should be merged after #1027.

- [x] Add the following notes:
  ```
  lambda_plus is increasing facilitation: too many daps, decrease facilitation!
  lambda_minus is increasing depression: too many daps, increase depression!
  if you see overlap, increase lambda_h a bit
  ```
- [x] eliminate the use of the "parameters" Python package

clinssen commented:
It seems as though the second line in the following snippet from the notebook causes an extreme memory consumption:
```py
connections = nest.GetConnections(synapse_model=...)
connections.set({'Wmin': connections.weight})
```
There are around 165000 connections.

pnbabu reviewed (changes requested):
Thanks for this excellent tutorial! I have a few minor comments:
- When you describe the different terms in the STDP rule in the text, could you refer to the equation by number? For example, the second term on the right-hand side in (1)...
- Any reason behind setting the number of threads to 1?
- Under "Generating code with NESTML", do we need the `try` and `except`?
	- Remove the following lines
```
"""module_name = "/tmp/nestml_target_68digoes/nestml_27fd8d9de14b408f9e6181b2f2310d41_module.so"
neuron_model_name = "iaf_psc_exp_nonlineardendrite_neuron_nestml__with_stdsp_synapse_nestml"
synapse_model_name = "stdsp_synapse_nestml__with_iaf_psc_exp_nonlineardendrite_neuron_nestml" """
```
- The fig just before `Experiment 2` is a bit hard to read. Can the three panels be a bit further apart?

clinssen commented on doc/tutorials/sequence_learning/iaf_psc_exp_nonlineardendrite_neuron.nestml:
```diff
@@ -0,0 +1,143 @@
+                    active_dendrite = false
+                    active_dendrite_readout = 0.
+
+                    # the following assignment to I_dend$ reproduces a bug in the original implementation
```
The model code reproduces the behaviour exactly of the original model, including the bug. I have updated the comment to reflect this.

clinssen commented on doc/tutorials/sequence_learning/iaf_psc_exp_nonlineardendrite_neuron.nestml:
```diff
@@ -0,0 +1,143 @@
+                    # starting dAP
+                    dAP_trace += 1 pA
+                    active_dendrite = true
+                    active_dendrite_readout = 1.
```
I have added a comment.

clinssen commented on doc/tutorials/sequence_learning/iaf_psc_exp_nonlineardendrite_neuron.nestml:
```diff
@@ -0,0 +1,143 @@
+        spike
+
+    onReceive(I_2):
+        I_dend$ += I_2 * ms * I_dend_incr * 1E6 # XXX factor 1E6?!
```
Fixed, "ms" should have been "s".

clinssen commented:
@pnbabu: much obliged for the review! I have cleaned up the notebook and the neuron model; this should address all your comments.

YounesBouhadjar commented:
I have created a PR #15 to your fork @clinssen with an update of the notebook.
Mainly minor things, except that I had to restructure a bit the third experiment,
for example, by adding new subtitles and changing the order of the subsections.

Remained to be done:

1. Add "taken from [1]" to the concerned figures.
2. We could add a legend describing the dashed and solid lines to the figure under experiment 1 showing the active dendrite simulations.
3. Could it be useful to provide the requirements of the packages to be installed? especially nest version? I saw also that a numpy<2.0.0 version is required.
4. The `nest.Install(module_name)` raises an error if called twice or more. A quick fix could be to do: try except

clinssen commented:
Thanks very much for the review! The latest changes should address all your comments.

> 1. Add "taken from [1]" to the concerned figures.

Done.

> 2. We could add a legend describing the dashed and solid lines to the figure under experiment 1 showing the active dendrite simulations.

Done.

> 3. Could it be useful to provide the requirements of the packages to be installed? especially nest version? I saw also that a numpy<2.0.0 version is required.

numpy is constrained to be >= 1.8.2. Only sympy is limited (to <1.11) but this should soon be amended (https://github.com/nest/ode-toolbox/pull/77#issuecomment-2491100824).

> 4. The `nest.Install(module_name)` raises an error if called twice or more. A quick fix could be to do: try except

This has been changed starting from NEST 3.7 (I noticed from the log messages in the notebook that you are using 3.6).

Please let me know if this addresses your comments; much obliged!

clinssen commented:
Thank you for the review!
